<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Run Canek on SingleCellExperiment objects • Canek</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Run Canek on SingleCellExperiment objects">
<meta property="og:description" content="Canek">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-52309538-4"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-52309538-4');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Canek</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Best_practices_thymus.html">Best practices on batch effects correction</a>
    </li>
    <li>
      <a href="../articles/SingleCellExperiment.html">Run Canek on SingleCellExperiment objects</a>
    </li>
    <li>
      <a href="../articles/seurat.html">Run Canek on Seurat objects</a>
    </li>
    <li>
      <a href="../articles/toy_example.html">Run Canek on a toy example</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/MartinLoza/Canek/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Run Canek on SingleCellExperiment objects</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/MartinLoza/Canek/blob/HEAD/vignettes/articles/SingleCellExperiment.Rmd" class="external-link"><code>vignettes/articles/SingleCellExperiment.Rmd</code></a></small>
      <div class="hidden name"><code>SingleCellExperiment.Rmd</code></div>

    </div>

    
    
<p>In this vignette, we demonstrate how to use <code>Canek</code> to
correct batch effects from <code>SingleCellExperiment</code> objects
following the recommendations from <a href="https://bioconductor.org/books/release/OSCA/" class="external-link">Bioconductor</a>
multi-sample analysis. The toy data sets used in this vignette are
included in the <code>Canek</code> R package.</p>
<div class="section level2">
<h2 id="init">Init<a class="anchor" aria-label="anchor" href="#init"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://martinloza.github.io/Canek/">Canek</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">batchelor</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="create-singlecellexperiment-objects">Create SingleCellExperiment objects<a class="anchor" aria-label="anchor" href="#create-singlecellexperiment-objects"></a>
</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Create independent SingleCellExperiment objects</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu">Canek</span><span class="fu">::</span><span class="va"><a href="../reference/SimBatches.html">SimBatches</a></span><span class="op">$</span><span class="va">batches</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">batch</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">counts</span> <span class="op">&lt;-</span> <span class="va">SimBatches</span><span class="op">$</span><span class="va">batches</span><span class="op">[[</span><span class="va">batch</span><span class="op">]</span><span class="op">]</span>     <span class="co">#get the counts</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>counts<span class="op">=</span><span class="va">counts</span><span class="op">)</span>,      <span class="co">#create the sce object</span></span>
<span>                       mainExpName <span class="op">=</span> <span class="st">"example_sce"</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Add batch labels</span></span>
<span><span class="va">names_batch</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"B1"</span>, <span class="st">"B2"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">names_batch</span></span>
<span><span class="va">sce</span><span class="op">$</span><span class="va">B1</span><span class="op">[[</span><span class="st">"batch"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"B1"</span></span>
<span><span class="va">sce</span><span class="op">$</span><span class="va">B2</span><span class="op">[[</span><span class="st">"batch"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"B2"</span></span>
<span></span>
<span><span class="co">#Add cell-type labels (included in Canek's package)</span></span>
<span><span class="va">celltypes</span> <span class="op">&lt;-</span> <span class="fu">Canek</span><span class="fu">::</span><span class="va"><a href="../reference/SimBatches.html">SimBatches</a></span><span class="op">$</span><span class="va">cell_types</span></span>
<span><span class="va">ct_B1_idx</span> <span class="op">&lt;-</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">sce</span><span class="op">$</span><span class="va">B1</span><span class="op">)</span></span>
<span><span class="va">ct_B2_idx</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">sce</span><span class="op">$</span><span class="va">B1</span><span class="op">)</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">celltypes</span><span class="op">)</span></span>
<span><span class="va">sce</span><span class="op">$</span><span class="va">B1</span><span class="op">[[</span><span class="st">"celltype"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">celltypes</span><span class="op">[</span><span class="va">ct_B1_idx</span><span class="op">]</span></span>
<span><span class="va">sce</span><span class="op">$</span><span class="va">B2</span><span class="op">[[</span><span class="st">"celltype"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">celltypes</span><span class="op">[</span><span class="va">ct_B2_idx</span><span class="op">]</span></span></code></pre></div>
<p>Let’s check the cells distribution across batches.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">sce</span><span class="op">$</span><span class="va">B1</span><span class="op">$</span><span class="va">batch</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  B1 </span></span>
<span><span class="co">#&gt; 631</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">sce</span><span class="op">$</span><span class="va">B2</span><span class="op">$</span><span class="va">batch</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  B2 </span></span>
<span><span class="co">#&gt; 948</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="pre-processing">Pre-processing<a class="anchor" aria-label="anchor" href="#pre-processing"></a>
</h2>
<p>In the pre-processing steps, we perform the normalization of counts
and the variance analysis of genes independently for each batch. Then,
we find common variable genes to use in downstream analyses.</p>
<div class="section level3">
<h3 id="normalization">Normalization<a class="anchor" aria-label="anchor" href="#normalization"></a>
</h3>
<p>We perform scaling normalization within each batch using the
<code>multiBatchNorm</code> function from <a href="https://bioconductor.org/packages/release/bioc/html/batchelor.html" class="external-link">batchelor</a>
Bioconductor package.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">batchelor</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/batchelor/man/multiBatchNorm.html" class="external-link">multiBatchNorm</a></span><span class="op">(</span><span class="va">sce</span><span class="op">[[</span><span class="st">"B1"</span><span class="op">]</span><span class="op">]</span>, <span class="va">sce</span><span class="op">[[</span><span class="st">"B2"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">names_batch</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="feature-selection">Feature selection<a class="anchor" aria-label="anchor" href="#feature-selection"></a>
</h3>
<p>We use the <code>combineVar</code> function from <a href="https://bioconductor.org/packages/release/bioc/html/scran.html" class="external-link">scran</a>
Bioconductor package. This function receives independent variance models
of genes for each of the batches. This allows us to select highly
variable genes while preserving within-batch differences.</p>
<p>We found 273 combined variable features.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Get independent features' variant models</span></span>
<span><span class="va">sce_var</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">sce</span>, FUN <span class="op">=</span> <span class="fu">scran</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html" class="external-link">modelGeneVar</a></span><span class="op">)</span> </span>
<span><span class="co">#Combine the independent variables</span></span>
<span><span class="va">combined_features</span> <span class="op">&lt;-</span> <span class="fu">scran</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scran/man/combineVar.html" class="external-link">combineVar</a></span><span class="op">(</span><span class="va">sce_var</span><span class="op">[[</span><span class="st">"B1"</span><span class="op">]</span><span class="op">]</span>, <span class="va">sce_var</span><span class="op">[[</span><span class="st">"B2"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#Select features using the `bio` statistics (see combineVar documentation for further details)</span></span>
<span><span class="va">combined_features</span> <span class="op">&lt;-</span> <span class="va">combined_features</span><span class="op">$</span><span class="va">bio</span> <span class="op">&gt;</span> <span class="fl">0</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">combined_features</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 273</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="visualization-before-batch-effects-correction">Visualization before batch effects correction<a class="anchor" aria-label="anchor" href="#visualization-before-batch-effects-correction"></a>
</h3>
<p>Then, we merge the two datasets and use the combined variable
features for the Principal Components Analysis.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">uncorrected</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">sce</span><span class="op">[[</span><span class="st">"B1"</span><span class="op">]</span><span class="op">]</span>, <span class="va">sce</span><span class="op">[[</span><span class="st">"B2"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">uncorrected</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">uncorrected</span><span class="op">[</span><span class="va">combined_features</span>,<span class="op">]</span>, scale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Let’s check out the elbow plot of the PCs.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pca_uncorrected</span> <span class="op">&lt;-</span> <span class="fu">SingleCellExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">uncorrected</span>, type <span class="op">=</span> <span class="st">"PCA"</span><span class="op">)</span></span>
<span><span class="va">variance</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">pca_uncorrected</span>, MARGIN <span class="op">=</span> <span class="fl">2</span>, FUN <span class="op">=</span> <span class="va">var</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>, y <span class="op">=</span> <span class="va">variance</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">30</span><span class="op">]</span>, xlab <span class="op">=</span> <span class="st">"Principal components"</span>, ylab <span class="op">=</span> <span class="st">"Variance"</span><span class="op">)</span></span></code></pre></div>
<p><img src="SingleCellExperiment_files/figure-html/unnamed-chunk-7-1.png" width="672" style="display: block; margin: auto;"></p>
<p>The variance stabilizes after the first seven PCs. In this test, we
will use 10 PCs to calculate the UMAP visualization, but feel free to
try other numbers.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">777</span><span class="op">)</span></span>
<span><span class="va">uncorrected</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">runUMAP</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">uncorrected</span>, dimred <span class="op">=</span> <span class="st">"PCA"</span>, pca <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<p>We can now visualize the uncorrected data by batch and cell-type
labels.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">uncorrected</span>, colour_by<span class="op">=</span><span class="st">"batch"</span><span class="op">)</span> </span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">uncorrected</span>, colour_by<span class="op">=</span><span class="st">"celltype"</span><span class="op">)</span> </span>
<span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span></span></code></pre></div>
<p><img src="SingleCellExperiment_files/figure-html/unnamed-chunk-9-1.png" width="672" style="display: block; margin: auto;"></p>
<p>We can observe that the cells labeled as <em>Cell Type 1</em> got
divided into two groups that correlated with batch labels. Let’s
minimize this batch differences with Canek.</p>
</div>
</div>
<div class="section level2">
<h2 id="correcting-batch-effects-with-canek">Correcting batch effects with Canek<a class="anchor" aria-label="anchor" href="#correcting-batch-effects-with-canek"></a>
</h2>
<p>Canek accepts a <code>SingleCellExperiments</code> object with a
<code>batch</code> identifier. We can pass a vector of variable features
to use in the integration.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">features</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">uncorrected</span><span class="op">)</span><span class="op">[</span><span class="va">combined_features</span><span class="op">]</span></span>
<span><span class="co">#RunCanek in a list of SingleCellExperiment objects</span></span>
<span><span class="co">#corrected &lt;- Canek::RunCanek(x = sce, features = features)</span></span>
<span><span class="co">#RunCanek in a single object with a batch identifier</span></span>
<span><span class="va">corrected</span> <span class="op">&lt;-</span> <span class="fu">Canek</span><span class="fu">::</span><span class="fu"><a href="../reference/RunCanek.html">RunCanek</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">uncorrected</span>, batches <span class="op">=</span> <span class="st">"batch"</span>, features <span class="op">=</span> <span class="va">features</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="visualization-after-batch-effects-correction-with-canek">Visualization after batch effects correction with Canek<a class="anchor" aria-label="anchor" href="#visualization-after-batch-effects-correction-with-canek"></a>
</h3>
<p>To perform PCA it’s important to use the corrected log counts. These
are saved in the assay <code>Canek</code> in the
<code>SingleCellExperiment</code> object and can be specified by
changing the <code>exprs_values</code> parameter.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">corrected</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">corrected</span>, scale <span class="op">=</span> <span class="cn">TRUE</span>, exprs_values <span class="op">=</span> <span class="st">"Canek"</span><span class="op">)</span></span></code></pre></div>
<p>Let’s check out the elbow plot after correction.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pca_corrected</span> <span class="op">&lt;-</span> <span class="fu">SingleCellExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">corrected</span>, type <span class="op">=</span> <span class="st">"PCA"</span><span class="op">)</span></span>
<span><span class="va">variance</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">pca_corrected</span>, MARGIN <span class="op">=</span> <span class="fl">2</span>, FUN <span class="op">=</span> <span class="va">var</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>, y <span class="op">=</span> <span class="va">variance</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">30</span><span class="op">]</span>, xlab <span class="op">=</span> <span class="st">"Principal components"</span>, ylab <span class="op">=</span> <span class="st">"Variance"</span><span class="op">)</span></span></code></pre></div>
<p><img src="SingleCellExperiment_files/figure-html/unnamed-chunk-12-1.png" width="672" style="display: block; margin: auto;"></p>
<p>After correction, the elbow plot has changed and 5 PCs will be enough
to calculate the UMAP visualization.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">777</span><span class="op">)</span></span>
<span><span class="va">corrected</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">runUMAP</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">corrected</span>, dimred <span class="op">=</span> <span class="st">"PCA"</span>, pca <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<p>We can now visualize the corrected data by batch and cell-type
labels. We can observe that after batch-correction with Canek, we
minimize the batch difference in cells from <em>Cell Type 1</em> while
preserving the other cell types.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">corrected</span>, colour_by<span class="op">=</span><span class="st">"batch"</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">corrected</span>, colour_by<span class="op">=</span><span class="st">"celltype"</span><span class="op">)</span></span>
<span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span></span></code></pre></div>
<p><img src="SingleCellExperiment_files/figure-html/unnamed-chunk-14-1.png" width="672" style="display: block; margin: auto;">
## Session info</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.3.2 (2023-10-31)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.3 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] patchwork_1.1.3             scran_1.28.2               </span></span>
<span><span class="co">#&gt;  [3] batchelor_1.16.0            scater_1.28.0              </span></span>
<span><span class="co">#&gt;  [5] ggplot2_3.4.4               scuttle_1.10.3             </span></span>
<span><span class="co">#&gt;  [7] SingleCellExperiment_1.22.0 SummarizedExperiment_1.30.2</span></span>
<span><span class="co">#&gt;  [9] Biobase_2.60.0              GenomicRanges_1.52.1       </span></span>
<span><span class="co">#&gt; [11] GenomeInfoDb_1.36.4         IRanges_2.34.1             </span></span>
<span><span class="co">#&gt; [13] S4Vectors_0.38.2            BiocGenerics_0.46.0        </span></span>
<span><span class="co">#&gt; [15] MatrixGenerics_1.12.3       matrixStats_1.0.0          </span></span>
<span><span class="co">#&gt; [17] Canek_0.2.4                </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] bitops_1.0-7              gridExtra_2.3            </span></span>
<span><span class="co">#&gt;   [3] rlang_1.1.1               magrittr_2.0.3           </span></span>
<span><span class="co">#&gt;   [5] compiler_4.3.2            flexmix_2.3-19           </span></span>
<span><span class="co">#&gt;   [7] DelayedMatrixStats_1.22.6 systemfonts_1.0.5        </span></span>
<span><span class="co">#&gt;   [9] vctrs_0.6.4               stringr_1.5.0            </span></span>
<span><span class="co">#&gt;  [11] pkgconfig_2.0.3           crayon_1.5.2             </span></span>
<span><span class="co">#&gt;  [13] fastmap_1.1.1             XVector_0.40.0           </span></span>
<span><span class="co">#&gt;  [15] labeling_0.4.3            utf8_1.2.4               </span></span>
<span><span class="co">#&gt;  [17] rmarkdown_2.25            ggbeeswarm_0.7.2         </span></span>
<span><span class="co">#&gt;  [19] ragg_1.2.6                numbers_0.8-5            </span></span>
<span><span class="co">#&gt;  [21] purrr_1.0.2               xfun_0.41                </span></span>
<span><span class="co">#&gt;  [23] modeltools_0.2-23         bluster_1.10.0           </span></span>
<span><span class="co">#&gt;  [25] zlibbioc_1.46.0           cachem_1.0.8             </span></span>
<span><span class="co">#&gt;  [27] beachmat_2.16.0           jsonlite_1.8.7           </span></span>
<span><span class="co">#&gt;  [29] highr_0.10                DelayedArray_0.26.7      </span></span>
<span><span class="co">#&gt;  [31] fpc_2.2-10                BiocParallel_1.34.2      </span></span>
<span><span class="co">#&gt;  [33] irlba_2.3.5.1             parallel_4.3.2           </span></span>
<span><span class="co">#&gt;  [35] prabclus_2.3-3            cluster_2.1.4            </span></span>
<span><span class="co">#&gt;  [37] R6_2.5.1                  bslib_0.5.1              </span></span>
<span><span class="co">#&gt;  [39] stringi_1.7.12            limma_3.56.2             </span></span>
<span><span class="co">#&gt;  [41] jquerylib_0.1.4           diptest_0.76-0           </span></span>
<span><span class="co">#&gt;  [43] Rcpp_1.0.11               knitr_1.45               </span></span>
<span><span class="co">#&gt;  [45] FNN_1.1.3.2               Matrix_1.6-1.1           </span></span>
<span><span class="co">#&gt;  [47] nnet_7.3-19               igraph_1.5.1             </span></span>
<span><span class="co">#&gt;  [49] tidyselect_1.2.0          viridis_0.6.4            </span></span>
<span><span class="co">#&gt;  [51] abind_1.4-5               yaml_2.3.7               </span></span>
<span><span class="co">#&gt;  [53] codetools_0.2-19          lattice_0.21-9           </span></span>
<span><span class="co">#&gt;  [55] tibble_3.2.1              withr_2.5.2              </span></span>
<span><span class="co">#&gt;  [57] evaluate_0.23             desc_1.4.2               </span></span>
<span><span class="co">#&gt;  [59] mclust_6.0.0              kernlab_0.9-32           </span></span>
<span><span class="co">#&gt;  [61] pillar_1.9.0              generics_0.1.3           </span></span>
<span><span class="co">#&gt;  [63] rprojroot_2.0.3           RCurl_1.98-1.13          </span></span>
<span><span class="co">#&gt;  [65] sparseMatrixStats_1.12.2  munsell_0.5.0            </span></span>
<span><span class="co">#&gt;  [67] scales_1.2.1              class_7.3-22             </span></span>
<span><span class="co">#&gt;  [69] glue_1.6.2                metapod_1.8.0            </span></span>
<span><span class="co">#&gt;  [71] tools_4.3.2               BiocNeighbors_1.18.0     </span></span>
<span><span class="co">#&gt;  [73] robustbase_0.99-0         ScaledMatrix_1.8.1       </span></span>
<span><span class="co">#&gt;  [75] locfit_1.5-9.8            fs_1.6.3                 </span></span>
<span><span class="co">#&gt;  [77] cowplot_1.1.1             grid_4.3.2               </span></span>
<span><span class="co">#&gt;  [79] edgeR_3.42.4              colorspace_2.1-0         </span></span>
<span><span class="co">#&gt;  [81] GenomeInfoDbData_1.2.10   beeswarm_0.4.0           </span></span>
<span><span class="co">#&gt;  [83] BiocSingular_1.16.0       vipor_0.4.5              </span></span>
<span><span class="co">#&gt;  [85] cli_3.6.1                 rsvd_1.0.5               </span></span>
<span><span class="co">#&gt;  [87] textshaping_0.3.7         fansi_1.0.5              </span></span>
<span><span class="co">#&gt;  [89] viridisLite_0.4.2         S4Arrays_1.0.6           </span></span>
<span><span class="co">#&gt;  [91] dplyr_1.1.3               uwot_0.1.16              </span></span>
<span><span class="co">#&gt;  [93] ResidualMatrix_1.10.0     gtable_0.3.4             </span></span>
<span><span class="co">#&gt;  [95] DEoptimR_1.1-3            sass_0.4.7               </span></span>
<span><span class="co">#&gt;  [97] digest_0.6.33             dqrng_0.3.1              </span></span>
<span><span class="co">#&gt;  [99] ggrepel_0.9.4             farver_2.1.1             </span></span>
<span><span class="co">#&gt; [101] memoise_2.0.1             htmltools_0.5.6.1        </span></span>
<span><span class="co">#&gt; [103] pkgdown_2.0.7             lifecycle_1.0.3          </span></span>
<span><span class="co">#&gt; [105] statmod_1.5.0             MASS_7.3-60</span></span></code></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Martin Loza, Diego Diez.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
